engine/execution/ingestion/engine.go

```diff
@@ -440,6 +454,31 @@ func (e *Engine) BlockProcessable(b *flow.Header) {
 	}
 }
 
+func (e *Engine) BlockFinalized(h *flow.Header) {
+
+	if e.pauseExecution {
+		return
+	}
+
+	stopHeight := e.stopAtHeight.Load()
+	// skip if no stop at height has been requested
+	if stopHeight == 0 {
+		return
+	}
+
+	// once finalization reached stop height we can be sure no other fork will be valid at this height,
+	// so it is safe to stop or crash now
+	if h.Height >= stopHeight {
+
+		if e.stopAtHeightCrash.Load() {
+			e.log.Fatal().Msgf("Crashing as finalization reached requested stop height %d", stopHeight)
```

Comment:
do we need to complete some finalization (like close\flush Badger etc.) before crashing here, to make sure data can still be intact, or is it really a don't care?

Reply:
Yes, we need finalization.
The thing is - we need to be sure that we executed all blocks up to height X, because we plan to use next version of EN/VN after restart, so it won't be possible to execute blocks below this height.
And we never know if there can be fork coming at any given height, unless this height is finalized.
Thats why we simply don't execute any blocks at or above X, but wait for X-1 to be finalized so we can be sure there are no possible forks.

Reply:
A block get finalized doesn't mean it has been executed. This could happen if an EN is falling behind.
We need to check whether the block at height `stopAtHeight - 1` has been executed, and only crash if it has been executed. 
If it has not executed yet, we should not change `e.pauseExecution = true`, but wait until the last block (`stopAtHeight - 1`) has been executed.